lib/scholar/neighbors/random_projection_forest.ex

```diff
@@ -0,0 +1,212 @@
+
+  @doc """
+  Grows a random projection forest.
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Examples
+
+    iex> key = Nx.Random.key(12)
+    iex> tensor = Nx.iota({5, 2})
+    iex> Scholar.Neighbors.RandomProjectionForest.fit(tensor, num_trees: 3, min_leaf_size: 2, key: key).indices
+    #Nx.Tensor<
+      u32[3][5]
+      [
+        [0, 1, 2, 3, 4],
+        [0, 1, 2, 3, 4],
+        [4, 3, 2, 1, 0]
+      ]
+    >
+  """
+  def fit(tensor, opts) do
+    if Nx.rank(tensor) != 2 do
+      raise ArgumentError,
+            """
+            expected input tensor to have shape {num_samples, num_features}, \
+            got tensor with shape: #{inspect(Nx.shape(tensor))}\
+            """
+    end
+
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+    min_leaf_size = opts[:min_leaf_size]
+    key = Keyword.get_lazy(opts, :key, fn -> Nx.Random.key(System.system_time()) end)
+
+    if min_leaf_size == 1 do
+      raise ArgumentError,
+            """
+            expected min_leaf_size to be at least 2, got 1
+            """
+    end
+
+    {size, dim} = Nx.shape(tensor)
+    {depth, leaf_size} = compute_depth_and_leaf_size(size, min_leaf_size, 0)
+
+    if depth == 0 do
+      raise ArgumentError,
+            """
+            expected num_samples to be at least twice \
+            min_leaf_size = #{inspect(min_leaf_size)}, got #{inspect(size)}
+            """
+    end
+
+    num_nodes = 2 ** depth - 1
+    num_trees = opts[:num_trees]
+
+    {hyperplanes, _key} =
+      Nx.Random.normal(key, type: to_float_type(tensor), shape: {num_trees, depth, dim})
+
+    medians = Nx.broadcast(:nan, {num_trees, num_nodes})
+    # TODO: Maybe rename acc to indices
+    acc = Nx.broadcast(Nx.u32(0), {num_trees, size})
+    root = 0
+    start_index = 0
+    indices = Nx.iota({1, size}, type: :u32) |> Nx.broadcast({num_trees, size})
+
+    {indices, medians} =
+      recur([{root, start_index, indices}], [], acc, leaf_size, 0, tensor, hyperplanes, medians)
+
+    %__MODULE__{
+      depth: depth,
+      leaf_size: leaf_size,
+      num_trees: num_trees,
+      indices: indices,
+      data: tensor,
+      hyperplanes: hyperplanes,
+      medians: medians
+    }
+  end
+
+  defp recur([], [], acc, _, _, _, _, medians) do
+    {acc, medians}
+  end
+
+  defp recur([], next, acc, leaf_size, level, tensor, hyperplanes, medians) do
+    recur(next, [], acc, leaf_size, level + 1, tensor, hyperplanes, medians)
+  end
+
+  defp recur(
+         [{node, start_index, indices} | rest],
+         next,
+         acc,
+         leaf_size,
+         level,
+         tensor,
+         hyperplanes,
+         medians
+       ) do
+    size = Nx.axis_size(indices, 1)
+
+    if size > leaf_size do
+      hyperplane = hyperplanes[[.., level]]
+
+      {median, left_indices, right_indices} =
+        Nx.Defn.jit_apply(&split(&1, &2, &3), [tensor, indices, hyperplane])
+
+      medians = Nx.put_slice(medians, [0, node], Nx.new_axis(median, 1))
```

Comment:
Move this inside `split`. We should have as many operations inside `defn` as possible.

Reply:
This is easy to do, but I don't see exactly why is it better given that `put_slice` is `Nx` function. Can I ask for an explanation?

Reply:
Nx in Elixir land needs to preserve the immutable semantics of the language. So every time you receive an output, the output will be fully copied into memory. In this case, with the two operations in the snippet, you are making a full copy of `median`, `left_indices`, `right_indices`, and `medians`. By moving this operation inside, at least you don't need to copy `median`.

This is exactly what moving the whole thing to a `defn` is beneficial. You don't have to worry about the copying of variables in a loop.

Reply:
I see, thanks.